library/alloc/src/rc.rs

```diff
@@ -1892,15 +1892,17 @@ impl<T> Weak<T> {
     pub fn as_ptr(&self) -> *const T {
         let ptr: *mut RcBox<T> = NonNull::as_ptr(self.ptr);
 
-        // SAFETY: we must offset the pointer manually, and said pointer may be
-        // a dangling weak (usize::MAX) if T is sized. data_offset is safe to call,
-        // because we know that a pointer to unsized T was derived from a real
-        // unsized T, as dangling weaks are only created for sized T. wrapping_offset
-        // is used so that we can use the same code path for the non-dangling
-        // unsized case and the potentially dangling sized case.
-        unsafe {
-            let offset = data_offset(ptr as *mut T);
-            set_data_ptr(ptr as *mut T, (ptr as *mut u8).wrapping_offset(offset))
```

Comment:
Does this mean `set_data_ptr` can be removed?

Reply:
Yeah, [`set_ptr_value`](https://github.com/rust-lang/rust/issues/75091) does the `set_data_ptr` operation in a potentially-exposable manner. `set_data_ptr` is still used in this file, and should be migrated to use `(*mut _)::set_ptr_value`; I felt it probably better to scope this PR to just the needed changes for `Weak::as_ptr`/`from_raw` though.

Reply:
> I felt it probably better to scope this PR to just the needed changes for Weak::as_ptr/from_raw though.

That makes sense, but using a mix of `set_data_ptr` and `set_ptr_value` makes this code harder to read. Maybe it's better to stick to `set_data_ptr` for now then.

Reply:
I went ahead and switched everything over to using `set_ptr_value`. `set_data_ptr` was marked `unsafe` which makes the impl a bit more clunky (and is why I originally used `set_ptr_value` for the adjusted impls).